t3d-mesh-partitioner/contrib/Ckit/avl/tree.3
.TH TREE 3 "22 Jan 1993"
.\" from .TH TREE 2 "23 June 1986"
.UC 4
.SH NAME
tree_init, tree_mung, tree_srch, tree_add, tree_delete, tree_trav
\- balanced binary tree routines
.SH SYNOPSIS
.nf
.B void
.B tree_init(tree)
.B void **tree;
.PP
.B int *
.B tree_srch(tree, compare, data)
.B void **tree;
.B int (*compare)();
.B void *data;
.PP
.B void
.B tree_add(tree, compare, data, del_uar)
.B void **tree;
.B int (*compare)();
.B void *data;
.B void (*del_uar)();
.PP
.B int
.B tree_delete(tree, compare, data, del_uar)
.B void **tree;
.B int (*compare)();
.B void *data;
.B void (*del_uar)();
.PP
.B int
.B tree_trav(tree, trav_uar)
.B void **tree;
.B int (*trav_uar)();
.PP
.B void
.B tree_mung(tree, del_uar)
.B void **tree;
.B void (*del_uar)();
.fi
.SH DESCRIPTION
These functions create and manipulate a balanced binary (AVL) tree.  Each node
of the tree contains the expected left & right subtree pointers, a short-int
balance indicator, and a pointer to the user-data.  On a 32-bit system, this
means an overhead of 4+4+2+4 bytes per node (or, on a RISC or otherwise
alignment-constrained system with implied padding, 4+4+4+4 bytes per node).
There is no key data type enforced by this package; a caller-supplied
compare routine is used to compare user-data blocks.
.PP
Balanced binary trees are very fast on searches and replacements, but have a
moderately high cost for additions and deletions.  If your application does a
lot more searches and replacements than it does additions and deletions, the
balanced (AVL) binary tree is a good choice for a data structure.
.PP
.I Tree_init
creates an empty tree and binds it to
.I tree
(which for this and all other routines in this package should be declared as
a pointer to void or int, and passed by reference), which can then be used by
other routines in this package.  Note that more than one
.I tree
variable can exist at once; thus multiple trees can be manipulated
simultaneously.
.PP
.I Tree_srch
searches a tree for a specific node and returns either
.I NULL
if no node was found, or the value of the user-data pointer if the node
was found.
.I compare
is the address of a function to compare two user-data blocks.  This routine
should work much the way 
.IR strcmp 2
does; in fact,
.I strcmp
could be used if the user-data was a null-terminated string.
.I data
is the address of a user-data block to be used by
.I compare
as the search criteria.  The tree is searched for a node where
.I compare
returns 0.
.PP
.I Tree_add
inserts or replaces a node in the specified tree.  The tree specified by
.I tree
is searched as in
.I tree_srch,
and if a node is found to match
.I data,
then the
.I del_uar
function is called with the address of the user-data block for the node
(this routine should deallocate any dynamic memory which is referenced
exclusively by the node); the user-data pointer for the node is then
replaced by the value of
.I data.
If no node is found to match, a new node is added (which may or may not
cause a transparent rebalance operation), with a user-data pointer equal to
.I data.
A rebalance may or may not occur, depending on where the node is added
and what the rest of the tree looks like.
.PP
.I Tree_delete
deletes a node from
.I tree.
A rebalance may or may not occur, depending on where the node is removed from
and what the rest of the tree looks like.
.I Tree_delete
returns TRUE if a node was deleted, FALSE otherwise.
.PP
.I Tree_trav
traverses all of
.I tree,
calling
.I trav_uar
with the address of each user-data block.  If
.I trav_uar
returns FALSE at any time,
.I tree_trav
will immediately return FALSE to its caller.  Otherwise all nodes will be 
reached and
.I tree_trav
will return TRUE.
.PP
.I Tree_mung
deletes every node in
.I tree,
calling
.I del_uar
with the user-data address from each node (see
.I tree_add
and
.I tree_delete
above).  The tree is left in the same state that
.I tree_init
leaves it in \- i.e., empty.
.SH AUTHOR
Paul Vixie, converted and augumented from Modula-2 examples in
.I Algorithms & Data Structures,
Niklaus Wirth, Prentice-Hall, ISBN 0-13-022005-1.


t3d-mesh-partitioner/contrib/Ckit/avl/tree.c
/* as_tree - tree library for as
 * vix 14dec85 [written]
 * vix 02feb86 [added tree balancing from wirth "a+ds=p" p. 220-221]
 * vix 06feb86 [added tree_mung()]
 * vix 20jun86 [added tree_delete per wirth a+ds (mod2 v.) p. 224]
 * vix 23jun86 [added delete uar to add for replaced nodes]
 * vix 22jan93 [revisited; uses RCS, ANSI, POSIX; has bug fixes]
 */


/* This program text was created by Paul Vixie using examples from the book:
 * "Algorithms & Data Structures," Niklaus Wirth, Prentice-Hall, 1986, ISBN
 * 0-13-022005-1.  This code and associated documentation is hereby placed
 * in the public domain.
 */


#ifndef LINT
static char RCSid[] = "$Id:";
#endif


/*#define		DEBUG	"tree"*/


#include <stdio.h>
#include <stdlib.h>
#include "vixie.h"
#include "tree.h"


#ifdef DEBUG
#define		MSG(msg)	printf("DEBUG: '%s'\n", msg);
#else
#define		MSG(msg)
#endif


#ifdef __STDC__
#define __P(x) x
#else
#define	__P(x) ()
#endif

static void	sprout	__P( (tree **, tree_t, int *, int (*)(), void (*)()) );
static int	delete	__P( (tree **, int (*)(), tree_t, void (*)(),
			      int *, int *) );
static void	del	__P( (tree **, int *, tree **, void (*)(), int *) );
static void	bal_L	__P( (tree **, int *) );
static void	bal_R	__P( (tree **, int *) );

#undef __P


void
tree_init(ppr_tree)
	tree	**ppr_tree;
{
	ENTER("tree_init")
	*ppr_tree = NULL;
	EXITV
}
	

tree_t
tree_srch(ppr_tree, pfi_compare, p_user)
	tree	**ppr_tree;
	int	(*pfi_compare)();
	tree_t	p_user;
{
	register int	i_comp;

	ENTER("tree_srch")

	if (*ppr_tree)
	{
		i_comp = (*pfi_compare)(p_user, (**ppr_tree).tree_p);
		if (i_comp > 0)
			EXIT(tree_srch(
				&(**ppr_tree).tree_r,
				pfi_compare,
				p_user
			))
		if (i_comp < 0)
			EXIT(tree_srch(
				&(**ppr_tree).tree_l,
				pfi_compare,
				p_user
			))

		/* not higher, not lower... this must be the one.
		 */
		EXIT((**ppr_tree).tree_p)
	}

	/* grounded. NOT found.
	 */
	EXIT(NULL)
}


void
tree_add(ppr_tree, pfi_compare, p_user, pfv_uar)
	tree	**ppr_tree;
	int	(*pfi_compare)();
	tree_t	p_user;
	void	(*pfv_uar)();
{
	int	i_balance = FALSE;

	ENTER("tree_add")
	sprout(ppr_tree, p_user, &i_balance, pfi_compare, pfv_uar);
	EXITV
}


int
tree_delete(ppr_p, pfi_compare, p_user, pfv_uar)
	tree	**ppr_p;
	int	(*pfi_compare)();
	tree_t	p_user;
	void	(*pfv_uar)();
{
	int	i_balance = FALSE,
		i_uar_called = FALSE;

	ENTER("tree_delete");
	EXIT(delete(ppr_p, pfi_compare, p_user, pfv_uar,
		    &i_balance, &i_uar_called))
}


int
tree_trav(ppr_tree, pfi_uar)
	tree	**ppr_tree;
	int	(*pfi_uar)();
{
	ENTER("tree_trav")

	if (!*ppr_tree)
		EXIT(TRUE)

	if (!tree_trav(&(**ppr_tree).tree_l, pfi_uar))
		EXIT(FALSE)
	if (!(*pfi_uar)((**ppr_tree).tree_p))
		EXIT(FALSE)
	if (!tree_trav(&(**ppr_tree).tree_r, pfi_uar))
		EXIT(FALSE)
	EXIT(TRUE)
}


void
tree_mung(ppr_tree, pfv_uar)
	tree	**ppr_tree;
	void	(*pfv_uar)();
{
	ENTER("tree_mung")
	if (*ppr_tree)
	{
		tree_mung(&(**ppr_tree).tree_l, pfv_uar);
		tree_mung(&(**ppr_tree).tree_r, pfv_uar);
		if (pfv_uar)
			(*pfv_uar)((**ppr_tree).tree_p);
		free(*ppr_tree);
		*ppr_tree = NULL;
	}
	EXITV
}


static void
sprout(ppr, p_data, pi_balance, pfi_compare, pfv_delete)
	tree	**ppr;
	tree_t	p_data;
	int	*pi_balance;
	int	(*pfi_compare)();
	void	(*pfv_delete)();
{
	tree	*p1, *p2;
	int	cmp;

	ENTER("sprout")

	/* are we grounded?  if so, add the node "here" and set the rebalance
	 * flag, then exit.
	 */
	if (!*ppr) {
		MSG("grounded. adding new node, setting h=true")
		*ppr = (tree *) malloc(sizeof(tree));
		(*ppr)->tree_l = NULL;
		(*ppr)->tree_r = NULL;
		(*ppr)->tree_b = 0;
		(*ppr)->tree_p = p_data;
		*pi_balance = TRUE;
		EXITV
	}

	/* compare the data using routine passed by caller.
	 */
	cmp = (*pfi_compare)(p_data, (*ppr)->tree_p);

	/* if LESS, prepare to move to the left.
	 */
	if (cmp < 0) {
		MSG("LESS. sprouting left.")
		sprout(&(*ppr)->tree_l, p_data, pi_balance,
			pfi_compare, pfv_delete);
		if (*pi_balance) {	/* left branch has grown longer */
			MSG("LESS: left branch has grown")
			switch ((*ppr)->tree_b)
			{
			case 1:	/* right branch WAS longer; balance is ok now */
				MSG("LESS: case 1.. balnce restored implicitly")
				(*ppr)->tree_b = 0;
				*pi_balance = FALSE;
				break;
			case 0:	/* balance WAS okay; now left branch longer */
				MSG("LESS: case 0.. balnce bad but still ok")
				(*ppr)->tree_b = -1;
				break;
			case -1:
				/* left branch was already too long. rebalnce */
				MSG("LESS: case -1: rebalancing")
				p1 = (*ppr)->tree_l;
				if (p1->tree_b == -1) {	/* LL */
					MSG("LESS: single LL")
					(*ppr)->tree_l = p1->tree_r;
					p1->tree_r = *ppr;
					(*ppr)->tree_b = 0;
					*ppr = p1;
				}
				else {			/* double LR */
					MSG("LESS: double LR")

					p2 = p1->tree_r;
					p1->tree_r = p2->tree_l;
					p2->tree_l = p1;

					(*ppr)->tree_l = p2->tree_r;
					p2->tree_r = *ppr;

					if (p2->tree_b == -1)
						(*ppr)->tree_b = 1;
					else
						(*ppr)->tree_b = 0;

					if (p2->tree_b == 1)
						p1->tree_b = -1;
					else
						p1->tree_b = 0;
					*ppr = p2;
				} /*else*/
				(*ppr)->tree_b = 0;
				*pi_balance = FALSE;
			} /*switch*/
		} /*if*/
		EXITV
	} /*if*/

	/* if MORE, prepare to move to the right.
	 */
	if (cmp > 0) {
		MSG("MORE: sprouting to the right")
		sprout(&(*ppr)->tree_r, p_data, pi_balance,
			pfi_compare, pfv_delete);
		if (*pi_balance) {	/* right branch has grown longer */
			MSG("MORE: right branch has grown")

			switch ((*ppr)->tree_b)
			{
			case -1:MSG("MORE: balance was off, fixed implicitly")
				(*ppr)->tree_b = 0;
				*pi_balance = FALSE;
				break;
			case 0:	MSG("MORE: balance was okay, now off but ok")
				(*ppr)->tree_b = 1;
				break;
			case 1:	MSG("MORE: balance was off, need to rebalance")
				p1 = (*ppr)->tree_r;
				if (p1->tree_b == 1) {	/* RR */
					MSG("MORE: single RR")
					(*ppr)->tree_r = p1->tree_l;
					p1->tree_l = *ppr;
					(*ppr)->tree_b = 0;
					*ppr = p1;
				}
				else {			/* double RL */
					MSG("MORE: double RL")

					p2 = p1->tree_l;
					p1->tree_l = p2->tree_r;
					p2->tree_r = p1;

					(*ppr)->tree_r = p2->tree_l;
					p2->tree_l = *ppr;

					if (p2->tree_b == 1)
						(*ppr)->tree_b = -1;
					else
						(*ppr)->tree_b = 0;

					if (p2->tree_b == -1)
						p1->tree_b = 1;
					else
						p1->tree_b = 0;

					*ppr = p2;
				} /*else*/
				(*ppr)->tree_b = 0;
				*pi_balance = FALSE;
			} /*switch*/
		} /*if*/
		EXITV
	} /*if*/

	/* not less, not more: this is the same key!  replace...
	 */
	MSG("I found it!  Replacing data value")
	*pi_balance = FALSE;
	if (pfv_delete)
		(*pfv_delete)((*ppr)->tree_p);
	(*ppr)->tree_p = p_data;
	EXITV
}


static int
delete(ppr_p, pfi_compare, p_user, pfv_uar, pi_balance, pi_uar_called)
	tree	**ppr_p;
	int	(*pfi_compare)();
	tree_t	p_user;
	void	(*pfv_uar)();
	int	*pi_balance;
	int	*pi_uar_called;
{
	tree	*pr_q;
	int	i_comp, i_ret;

	ENTER("delete")

	if (*ppr_p == NULL) {
		MSG("key not in tree")
		EXIT(FALSE)
	}

	i_comp = (*pfi_compare)((*ppr_p)->tree_p, p_user);
	if (i_comp > 0) {
		MSG("too high - scan left")
		i_ret = delete(&(*ppr_p)->tree_l, pfi_compare, p_user, pfv_uar,
						pi_balance, pi_uar_called);
		if (*pi_balance)
			bal_L(ppr_p, pi_balance);
	}
	else if (i_comp < 0) {
		MSG("too low - scan right")
		i_ret = delete(&(*ppr_p)->tree_r, pfi_compare, p_user, pfv_uar,
						pi_balance, pi_uar_called);
		if (*pi_balance)
			bal_R(ppr_p, pi_balance);
	}
	else {
		MSG("equal")
		pr_q = *ppr_p;
		if (pr_q->tree_r == NULL) {
			MSG("right subtree null")
			*ppr_p = pr_q->tree_l;
			*pi_balance = TRUE;
		}
		else if (pr_q->tree_l == NULL) {
			MSG("right subtree non-null, left subtree null")
			*ppr_p = pr_q->tree_r;
			*pi_balance = TRUE;
		}
		else {
			MSG("neither subtree null")
			del(&pr_q->tree_l, pi_balance, &pr_q, pfv_uar,
								pi_uar_called);
			if (*pi_balance)
				bal_L(ppr_p, pi_balance);
		}
		if (!*pi_uar_called && *pfv_uar)
			(*pfv_uar)(pr_q->tree_p);
		free(pr_q);	/* thanks to wuth@castrov.cuc.ab.ca */
		i_ret = TRUE;
	}
	EXIT(i_ret)
}


static void
del(ppr_r, pi_balance, ppr_q, pfv_uar, pi_uar_called)
	tree	**ppr_r;
	int	*pi_balance;
	tree	**ppr_q;
	void	(*pfv_uar)();
	int	*pi_uar_called;
{
	ENTER("del")

	if ((*ppr_r)->tree_r != NULL) {
		del(&(*ppr_r)->tree_r, pi_balance, ppr_q,
		    pfv_uar, pi_uar_called);
		if (*pi_balance)
			bal_R(ppr_r, pi_balance);
	} else {
		if (pfv_uar)
			(*pfv_uar)((*ppr_q)->tree_p);
		*pi_uar_called = TRUE;
		(*ppr_q)->tree_p = (*ppr_r)->tree_p;
		*ppr_q = *ppr_r;
		*ppr_r = (*ppr_r)->tree_l;
		*pi_balance = TRUE;
	}

	EXITV
}


static void
bal_L(ppr_p, pi_balance)
	tree	**ppr_p;
	int	*pi_balance;
{
	tree	*p1, *p2;
	int	b1, b2;

	ENTER("bal_L")
	MSG("left branch has shrunk")

	switch ((*ppr_p)->tree_b)
	{
	case -1: MSG("was imbalanced, fixed implicitly")
		(*ppr_p)->tree_b = 0;
		break;
	case 0:	MSG("was okay, is now one off")
		(*ppr_p)->tree_b = 1;
		*pi_balance = FALSE;
		break;
	case 1:	MSG("was already off, this is too much")
		p1 = (*ppr_p)->tree_r;
		b1 = p1->tree_b;
		if (b1 >= 0) {
			MSG("single RR")
			(*ppr_p)->tree_r = p1->tree_l;
			p1->tree_l = *ppr_p;
			if (b1 == 0) {
				MSG("b1 == 0")
				(*ppr_p)->tree_b = 1;
				p1->tree_b = -1;
				*pi_balance = FALSE;
			} else {
				MSG("b1 != 0")
				(*ppr_p)->tree_b = 0;
				p1->tree_b = 0;
			}
			*ppr_p = p1;
		} else {
			MSG("double RL")
			p2 = p1->tree_l;
			b2 = p2->tree_b;
			p1->tree_l = p2->tree_r;
			p2->tree_r = p1;
			(*ppr_p)->tree_r = p2->tree_l;
			p2->tree_l = *ppr_p;
			if (b2 == 1)
				(*ppr_p)->tree_b = -1;
			else
				(*ppr_p)->tree_b = 0;
			if (b2 == -1)
				p1->tree_b = 1;
			else
				p1->tree_b = 0;
			*ppr_p = p2;
			p2->tree_b = 0;
		}
	}
	EXITV
}


static void
bal_R(ppr_p, pi_balance)
	tree	**ppr_p;
	int	*pi_balance;
{
	tree	*p1, *p2;
	int	b1, b2;

	ENTER("bal_R")
	MSG("right branch has shrunk")
	switch ((*ppr_p)->tree_b)
	{
	case 1:	MSG("was imbalanced, fixed implicitly")
		(*ppr_p)->tree_b = 0;
		break;
	case 0:	MSG("was okay, is now one off")
		(*ppr_p)->tree_b = -1;
		*pi_balance = FALSE;
		break;
	case -1: MSG("was already off, this is too much")
		p1 = (*ppr_p)->tree_l;
		b1 = p1->tree_b;
		if (b1 <= 0) {
			MSG("single LL")
			(*ppr_p)->tree_l = p1->tree_r;
			p1->tree_r = *ppr_p;
			if (b1 == 0) {
				MSG("b1 == 0")
				(*ppr_p)->tree_b = -1;
				p1->tree_b = 1;
				*pi_balance = FALSE;
			} else {
				MSG("b1 != 0")
				(*ppr_p)->tree_b = 0;
				p1->tree_b = 0;
			}
			*ppr_p = p1;
		} else {
			MSG("double LR")
			p2 = p1->tree_r;
			b2 = p2->tree_b;
			p1->tree_r = p2->tree_l;
			p2->tree_l = p1;
			(*ppr_p)->tree_l = p2->tree_r;
			p2->tree_r = *ppr_p;
			if (b2 == -1)
				(*ppr_p)->tree_b = 1;
			else
				(*ppr_p)->tree_b = 0;
			if (b2 == 1)
				p1->tree_b = -1;
			else
				p1->tree_b = 0;
			*ppr_p = p2;
			p2->tree_b = 0;
		}
	}
	EXITV
}
